### ISSUE TITLE
Kim CNN: Non-deterministic results

### ISSUE THREAD

**Issue by rosequ** [castorini/Castor#45]
### Configuration:
__Dataset:__ SST-1
__Mode:__ multichannel
This was run on dragon with GPU: 

__Accuracy:__
run1: 45.47061422889969
run2: 43.923994697304465

Something worth looking at:
https://discuss.pytorch.org/t/non-reproducible-result-with-gpu/1831

**Comment by Impavidity** [comment]
I will look into it for better solution.

**Comment by rosequ** [comment]
@Impavidity I will try disabling `cudann` first and see if it works

**Comment by rosequ** [comment]
I repeated the experiments on TREC dataset because SST-1 takes longer to train with `cudann` disabled. 
For TREC, I get the same numbers regardless of whether `cudann` is enabled or not.

**Comment by Impavidity** [comment]
I tried to install the pytorch from source with latest version and re-run the Kim-CNN, the training loss and accuracy is deterministic now. But the training time is much longer (takes hours) than non-deterministic version (<20min). Will update experiment results. 
```
torch.backends.cudnn.deterministic = True
```
is used for deterministic results.

**Comment by snapbug** [comment]
Out of interest how does the training time with cudnn completely disabled compare to training with cudnn deterministic?

**Comment by Impavidity** [comment]
(deterministic : disable) = (6~7) : 1 in my case.

**Comment by Impavidity** [comment]
Deterministic Algorithm for CNN. 
| Dev Accuracy on SST-1 |     rand      |    static    |   non-static  |  multichannel |
|:--------------------------:|:-----------:|:-----------:|:-------------:|:---------------:|
| My-Implementation      | 42.597639| 48.773842| 48.864668   | 

| Test Accuracy on SST-1|      rand      |    static    |    non-static |  multichannel |
|:--------------------------:|:-----------:|:-----------:|:-------------:|:---------------:|
| Kim-Implementation    | 45.0            | 45.5        | 48.0             | 47.4                 |
| My- Implementation    | 39.683258  | 45.972851| 48.914027|  47.330317       |

**Comment by snapbug** [comment]
So if I'm reading that speed difference correctly -- 

non deterministic cudnn is ~20m
disabled cudnn is ~1h (and also reproducible)
deterministic cudnn is ~6h

Interesting.

**Comment by Impavidity** [comment]
deterministic cudnn is ~1h and disabled cudnn is ~6h. I do not look into the speed of non deterministic cudnn and deterministic cudnn.

**Comment by lintool** [comment]
@Impavidity the results above are different from what's in the README. Can you make sure the latest figures are updated in the README?

Can you please document all the different variants (rmsprop, adagrad, etc.) and the cudnn settings, as well as the running times?
